File: cc/029.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
  public:
    int MoreThanHalfNum_Solution(vector<int> numbers)
    {
        auto result = 0;
        if (!numbers.empty())
        {
            auto count = 0;
            for (auto it = numbers.begin(); it != numbers.end(); it++)
            {
                result = count == 0 ? *it : result;
                count = result == *it ? count + 1 : count - 1;
            }
            count = 0;
            for (auto it = numbers.begin(); it != numbers.end(); it++)
                count = *it == result ? count + 1 : count;
            result = count * 2 > (int)numbers.size() ? result : 0;
        }
        return result;
    }
};

int main(int argc, char const *argv[])
{
    vector<int> vct0 = {1, 2, 1, 3, 1};
    cout << (new Solution())->MoreThanHalfNum_Solution(vct0) << endl; // 1
    vector<int> vct1 = {1, 2, 2, 3, 1};
    cout << (new Solution())->MoreThanHalfNum_Solution(vct1) << endl; // 0
    return 0;
}

File: python/029.py
# -*- coding: utf-8 -*-
# /usr/bin/python3
# Creator: Hemajun
# Description: 数组中有一个数字出现的次数超过数组长度的一半，请找出这个数字。例如输入一个长度为9的数组{1,2,3,2,2,2,5,4,2}。由于数字2在数组中出现了5次，超过数组长度的一半，因此输出2。如果不存在则输出0。


class Solution:
    def MoreThanHalfNum_Solution(self, numbers):
        result = 0
        if len(numbers) > 0:
            count = 0
            for num in numbers:
                result = num if count == 0 else result
                count = count + 1 if result == num else count - 1
            count = 0
            for num in numbers:
                count = count + 1 if result == num else count
            result = result if 2 * count > len(numbers) else 0
        return result
